File: regular_expression_matching/Regular_Expression_Matching.cpp
class Regular_Expression_Matching {
public:
    bool isMatch(string s, string p) {
        if (s.empty() && p.empty()) {
            return true;
        }
        
        bool dp_table[s.size() + 1][p.size() + 1];
        for (int i = 0; i <= s.size(); i++) {
            for (int j = 0; j <= p.size(); j++) {
                dp_table[i][j] = false;
            }
        }

        dp_table[0][0] = true;
        for (int i = 2; i <= p.size(); i++) {
            dp_table[0][i] = p[i - 1] == '*' && dp_table[0][i - 2];
        }
        
        for (int sptr = 1; sptr <= s.size(); sptr++) {
            for (int pptr = 1; pptr <= p.size(); pptr++) {
                if (isalpha(p[pptr - 1])) {
                    dp_table[sptr][pptr] = s[sptr - 1] == p[pptr - 1] && dp_table[sptr - 1][pptr - 1];
                    continue;
                }
                if (p[pptr - 1] == '.') {
                    dp_table[sptr][pptr] = dp_table[sptr - 1][pptr - 1];
                    continue;
                }
                
                dp_table[sptr][pptr] = dp_table[sptr][pptr - 2] || ((s[sptr - 1] == p[pptr - 2] || p[pptr - 2] == '.') && dp_table[sptr - 1][pptr]);
            }
        }
        
        return dp_table[s.size()][p.size()];
    }
};

File: regular_expression_matching/RegularExpressionMatching.java
class RegularExpressionMatching {
    public boolean isMatch(String s, String p) {
        if (s == null && p == null) {
            return true;
        }
        if (s == null || p == null) {
            return false;
        }
        
        boolean[][] dpTable = new boolean[s.length() + 1][p.length() + 1];
        dpTable[0][0] = true;
        for (int i = 2; i <= p.length(); i++) {
            dpTable[0][i] = p.charAt(i - 1) == '*' && dpTable[0][i - 2];
        }
        
        for (int sptr = 1; sptr < dpTable.length; sptr++) {
            for (int pptr = 1; pptr < dpTable[sptr].length; pptr++) {
                char sch = s.charAt(sptr - 1);
                char pch = p.charAt(pptr - 1);
                if (Character.isLetter(pch)) {
                    dpTable[sptr][pptr] = (sch == pch) && dpTable[sptr - 1][pptr - 1];
                    continue;
                }
                if (pch == '.') {
                    dpTable[sptr][pptr] = dpTable[sptr - 1][pptr - 1];
                    continue;
                }
                
                char ppch = p.charAt(pptr - 2);
                dpTable[sptr][pptr] = dpTable[sptr][pptr - 2] || ((sch == ppch || ppch == '.') && dpTable[sptr - 1][pptr]);
            }
        }
        
        return dpTable[s.length()][p.length()];
    }
}